src/Framework/Models/Model.php

```diff
@@ -0,0 +1,337 @@
+<?php
+
+namespace Give\Framework\Models;
+
+use Give\Framework\Exceptions\Primitives\InvalidArgumentException;
+use Give\Framework\Models\ValueObjects\Relationship;
+use Give\Framework\Support\Contracts\Arrayable;
+use RuntimeException;
+
+/**
+ * @unreleased
+ */
+abstract class Model implements Arrayable
+{
+
+    /**
+     * The model's attributes.
+     *
+     * @var array
+     */
+    protected $attributes = [];
+
+    /**
+     * The model attribute's original state.
+     *
+     * @var array
+     */
+    protected $original = [];
+
+    /**
+     * The model properties assigned to their types
+     *
+     * @var array
+     */
+    protected $properties = [];
+
+    /**
+     * The model relationships assigned to their relationship types
+     *
+     * @var array
+     */
+    protected $relationships = [];
+
+    /**
+     * Create a new model instance.
+     *
+     * @unreleased
+     *
+     * @param  array  $attributes
+     * @return void
+     */
+    public function __construct(array $attributes = [])
+    {
+        $this->attributes = $attributes;
+
+        $this->syncOriginal();
+
+        $this->fill($attributes);
+    }
+
+    /**
+     * Sync the original attributes with the current.
+     *
+     * @unreleased
+     *
+     * @return $this
+     */
+    protected function syncOriginal()
+    {
+        $this->original = $this->attributes;
+
+        return $this;
+    }
+
+    /**
+     * Get the model's original attribute values.
+     *
+     * @unreleased
+     *
+     * @param  string|null  $key
+     * @return mixed|array
+     */
+    public function getOriginal($key = null)
+    {
+        return $key ? $this->original[$key] : $this->original;
+    }
+
+    /**
+     * Determine if a given attribute is dirty.
+     *
+     * @unreleased
+     *
+     * @param  string  $attribute
+     * @return bool
+     */
+    public function isDirty($attribute)
+    {
+        return array_key_exists($attribute, $this->getDirty());
+    }
```

Comment:
Let's make `$attribute` optional; if it's null then it checks if _any_ value is dirty.

Also, let's add an inverted method: `isClean` — which can just call this one and reverse the boolean.

Reply:
all set 88c4d73f